sympy/solvers/diophantine.py

```diff
@@ -532,35 +484,27 @@ def _diop_linear(var, coeff, param):
 
     This method is generalised for many variables, below.
 
-    """
+    '''
 
     solutions = []
     no_solution = tuple([None] * len(var))
     for i in range(len(B)):
         tot_x, tot_y = 0, 0
 
-        if isinstance(c, Add):
-            # example: 5 + t_0 + 3*t_1
-            args = c.args
-        else: # c is a Mul, a Symbol, or an Integer
-            args = [c]
-
-        for j in range(len(args)):
-            if isinstance(args[j], Mul):
-                # example: 3*t_1 -> k = 3
-                k = args[j].as_two_terms()[0]
-                param_index = params.index(args[j].as_two_terms()[1]) + 1
-            elif isinstance(args[j], Symbol):
-                # example: t_0 -> k = 1
-                k = 1
-                param_index = params.index(args[j]) + 1
-            else: #args[j] is an Integer
+        for j, arg in enumerate(Add.make_args(c)):
+            k, p = arg.as_coeff_Mul()
+            if p.is_Integer:   # arg is an Integer
                 # example: 5 -> k = 5
-                k = args[j]
                 param_index = 0
+                if not p:
```

Comment:
for this reason, `as_coeff_Mul` should probably always return (0, 0) for 0. Check codebase to see how it's used.